FireRender.Maya.Src/frWrap.h

```diff
@@ -1199,19 +1199,49 @@ namespace frw
 
 	class Image : public Object
 	{
-		DECLARE_OBJECT_NO_DATA(Image, Object);
+		DECLARE_OBJECT(Image, Object);
+
+		class Data : public Object::Data
+		{
+			DECLARE_OBJECT_DATA
+		public:
+			virtual ~Data()
+			{
+				for (auto it = m_udimsMap.begin(); it != m_udimsMap.end(); ++it)
+				{
+					rprImageSetUDIM(Handle(), it->first, nullptr);
+				}
+
+				m_udimsMap.clear();
+			}
+			
+			std::map<rpr_uint, Image> m_udimsMap;
+		};
+
+
 	public:
 		explicit Image(rpr_image h, const Context& context) : Object(h, context, true, new Data()) {}
 
 		Image(Context context, float r, float g, float b);
 		Image(Context context, const rpr_image_format& format, const rpr_image_desc& image_desc, const void* data);
+
+		// create empty image (used for UDIM creation of master image
+		Image(Context context, const rpr_image_format& format);
 		Image(Context context, const char * filename);
 		void SetGamma(float gamma)
 		{
 			rpr_int res = rprImageSetGamma(Handle(), gamma);
 			checkStatus(res);
 		}
 
+		void SetUDIM(rpr_uint tileIndex, frw::Image image)
+		{
+			rpr_int res = rprImageSetUDIM(Handle(), tileIndex, image.Handle());
+			checkStatus(res);
+
+			data().m_udimsMap[tileIndex] = image;
```

Comment:
Maybe add some check that tileIndex is unique?

Reply:
No, because technically you can overwrite image by using the same tile index